### PR TITLE
User auth endpoint

### LINKED ISSUE #2: Implement `POST /karen/v1/users/auth` API
Create an API to authenticate a user.

The request body is expected to look like the body outlined below.
```
{
  "email": string,
  "password": string
}
```
The `email` will be used to confirm the existence of the user and retrieve the hashed password and salt for the user. The salt will then be used to hash the `password` in the request body and the two hashed passwords will be compared.

The possible responses are:
* 200 OK if the passwords match and the body must look like the JSON below.
```
{
  "id": int,
  "name": string,
  "email": string
}
```
* 400 Bad Request if the request body is malformed or syntactically incorrect
* 404 Not Found if the user cannot be found
* 500 Internal Server Error in any other case with a string describing the error

### PR DESCRIPTION
**Pull request by mRileyMacK**
Resolves #2

### CODE REVIEW

File: handlers/handler.go [hunk 1]
```diff
@@ -26,6 +26,45 @@ func internalServerError(w http.ResponseWriter, err error) {
...
+	if err := parseJSON(w, r.Body, reqUser); err != nil {
+		return
+	}
+	if reqUser.Email == "" || reqUser.Password == "" {
+		errMsg := "Request body is missing field(s)"
+		log.Println(errMsg)
+		http.Error(w, errMsg, http.StatusBadRequest)
+		return
+	}
+	user := &models.User{}
```
Comment (Thao-Tran) [inline handlers/handler.go]: Don't need this since you get it from the method anyway.
```suggestion
```

File: handlers/handler.go [hunk 2]
```diff
@@ -26,6 +26,45 @@ func internalServerError(w http.ResponseWriter, err error) {
...
+	if err != nil {
+		mySQLErr, ok := err.(*mysql.MySQLError)
+		if ok && mySQLErr.Number == 1065 {
+			errMsg := fmt.Sprintf("User with email %s was not found", reqUser.Email)
+			log.Println(errMsg)
+			http.Error(w, errMsg, http.StatusNotFound)
+		} else if err.Error() == "password incorrect" {
+			errMsg := fmt.Sprintf("Password incorrect")
+			log.Println(errMsg)
+			http.Error(w, errMsg, http.StatusUnauthorized)
```
Comment (Thao-Tran) [inline handlers/handler.go]: The capital `Password` isn't super necessary, so you can just use `err.Error()`.
```suggestion
			log.Println(err.Error())
			http.Error(w, err.Error(), http.StatusUnauthorized)
```

File: handlers/handler.go [hunk 3]
```diff
@@ -26,6 +26,45 @@ func internalServerError(w http.ResponseWriter, err error) {
...
+	}
+	reqUser.ID = user.ID
+	reqUser.Name = user.Name
+	reqUser.Password = ""
+	location := fmt.Sprintf("%s/%d", r.URL.Path, user.ID)
+	w.Header().Add("Location", location)
+	w.Header().Add("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(reqUser)
+
```
Comment (Thao-Tran) [inline handlers/handler.go]: Remove unnecessary whitespace.
```suggestion
```

File: handlers/handler.go [hunk 4]
```diff
@@ -26,6 +26,45 @@ func internalServerError(w http.ResponseWriter, err error) {
...
+		} else {
+			internalServerError(w, err)
+		}
+		return
+	}
+	reqUser.ID = user.ID
+	reqUser.Name = user.Name
+	reqUser.Password = ""
+	location := fmt.Sprintf("%s/%d", r.URL.Path, user.ID)
+	w.Header().Add("Location", location)
```
Comment (Thao-Tran) [inline handlers/handler.go]: Don't have to set a `Location` header in this case, since nothing is being created.
```suggestion
```

File: models/user.go [hunk 5]
```diff
@@ -20,6 +21,22 @@ const (
...
+	userFromDB := &User{}
+	queryString := fmt.Sprintf("SELECT ID, Name, Password FROM %s WHERE EMAIL=?", usersTable)
+	if err := db.QueryRow(queryString, user.Email).Scan(&(userFromDB.ID),
+		&(userFromDB.Name), &(userFromDB.Password)); err != nil {
+		return nil, err
+	}
+	err := checkPasswordHash(user.Password, userFromDB.Password)
+	if err != nil {
+		err := errors.New("password incorrect")
+		return nil, err
```
Comment (Thao-Tran) [inline models/user.go]: Just return the new error straight-up to make it cleaner.
```suggestion
		return nil, errors.New("password incorrect")
```

File: handlers/handler.go (hunk 4 above)
Comment (mRileyMacK) [reply handlers/handler.go]: ahh duh doi!

File: models/user.go [hunk 6]
```diff
@@ -20,6 +21,21 @@ const (
 	usersTable string = "users"
 )
 
+//Authenticates the user's credentials and returns the user
+func (db *DB) CheckUser(user *User) (*User, error) {
+	userFromDB := &User{}
+	queryString := fmt.Sprintf("SELECT ID, Name, Password FROM %s WHERE EMAIL=?", usersTable)
+	if err := db.QueryRow(queryString, user.Email).Scan(&(userFromDB.ID),
```
Comment (igorveselinovic) [inline models/user.go]: Might be more readable to break this up into two lines, like:
```
err := db.QueryRow()
if err != nil {
```

Comment (mRileyMacK) [reply models/user.go]: oh true, I'll fix that rn

Comment (mRileyMacK) [reply models/user.go]: done
